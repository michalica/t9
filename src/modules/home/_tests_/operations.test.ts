import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import fetchMock from 'fetch-mock';
import createDebounce from 'redux-debounced';
import {
    fetchT9api,
    doRequest
}
    from "../operations";

export const createStore = (defaultState: any) => {
    const middlewares = [createDebounce(), thunk];
    const mock = configureMockStore(middlewares);
    return mock(defaultState);
};

const defaultState = {
    modules: {
        home: {
            letters: [],
            pagination: 0,
            isLoading: false,
            hasError: false,
            form: {
                validationError: false,
                value: '',
            }
        },
    }
};
describe('testing operation fetT9Api', () => {
    afterEach(() => {
        fetchMock.reset();
        fetchMock.restore();
    });

    it('fetch with string', () => {
        fetchMock
            .getOnce('http://localhost:3001/num?num=1234&page=0', {
                body: {
                    letters: ['abc', 'def']
                },
                headers: {'content-type': 'application/json'}
            })
        const store = createStore(defaultState);
        return store.dispatch(fetchT9api('1234'))
            .then(() => {
                expect(store.getActions()).toMatchSnapshot();
            });
    });
    it('fetch with empty string', () => {
        fetchMock
            .getOnce('http://localhost:3001/num?num=1234&page=0', {
                body: {
                    letters: ['abc', 'def']
                },
                headers: {'content-type': 'application/json'}
            })
        const store = createStore(defaultState);
        store.dispatch(fetchT9api(''));
        expect(store.getActions()).toMatchSnapshot();
    });

    it('fetch with undefined', () => {
        fetchMock
            .getOnce('http://localhost:3001/num?num=1234&page=0', {
                body: {
                    letters: ['abc', 'def']
                },
                headers: {'content-type': 'application/json'}
            })
        const store = createStore(defaultState);
        store.dispatch(fetchT9api(null));
        expect(store.getActions()).toMatchSnapshot();
    });
});

describe('testing operations doReqeuest', () => {
    afterEach(() => {
        fetchMock.reset();
        fetchMock.restore();
    });

    it('doRequest with string, replace letters', () => {
        fetchMock
            .getOnce('http://localhost:3001/num?num=1234&page=0', {
                body: {
                    letters: ['abc', 'def']
                },
                headers: {'content-type': 'application/json'}
            })
        const store = createStore(defaultState);
        return store.dispatch(doRequest('1234'))
            .then(() => {
                expect(store.getActions()).toMatchSnapshot();
            });
    });

    it('doRequest with string, update letters', () => {
        fetchMock
            .getOnce('http://localhost:3001/num?num=1234&page=0', {
                body: {
                    letters: ['abc', 'def']
                },
                headers: {'content-type': 'application/json'}
            })
        const store = createStore(defaultState);
        return store.dispatch(doRequest('1234', true))
            .then(() => {
                expect(store.getActions()).toMatchSnapshot();
            });
    });


    it('doRequest with undefined, error expected', () => {
        fetchMock
            .getOnce('http://localhost:3001/num?num=1234&page=0', {
                body: {
                    letters: ['abc', 'def']
                },
                headers: {'content-type': 'application/json'}
            })
        const store = createStore(defaultState);
        return store.dispatch(doRequest(null ))
            .then(() => {
                expect(store.getActions()).toMatchSnapshot();
            });
    });
});

