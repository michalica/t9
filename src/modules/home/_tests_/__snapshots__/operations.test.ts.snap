// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing operation fetT9Api fetch with empty string 1`] = `
Array [
  Object {
    "type": "RESET_PAGE",
  },
  Object {
    "payload": "",
    "type": "SET_FORM_VALUE",
  },
  Object {
    "payload": false,
    "type": "VALIDATION_ERROR",
  },
  Object {
    "type": "REMOVE_LETTERS",
  },
]
`;

exports[`testing operation fetT9Api fetch with string 1`] = `
Array [
  Object {
    "type": "RESET_PAGE",
  },
  Object {
    "payload": "1234",
    "type": "SET_FORM_VALUE",
  },
  Object {
    "payload": false,
    "type": "VALIDATION_ERROR",
  },
  Object {
    "payload": true,
    "type": "IS_LOADING",
  },
  Object {
    "payload": Object {
      "letters": Array [
        "abc",
        "def",
      ],
    },
    "type": "SET_LETTERS",
  },
  Object {
    "type": "INCREMENT_PAGE",
  },
  Object {
    "payload": false,
    "type": "IS_LOADING",
  },
]
`;

exports[`testing operation fetT9Api fetch with undefined 1`] = `
Array [
  Object {
    "type": "RESET_PAGE",
  },
  Object {
    "payload": null,
    "type": "SET_FORM_VALUE",
  },
  Object {
    "payload": true,
    "type": "VALIDATION_ERROR",
  },
]
`;

exports[`testing operations doReqeuest doRequest with string, replace letters 1`] = `
Array [
  Object {
    "payload": true,
    "type": "IS_LOADING",
  },
  Object {
    "payload": Object {
      "letters": Array [
        "abc",
        "def",
      ],
    },
    "type": "SET_LETTERS",
  },
  Object {
    "type": "INCREMENT_PAGE",
  },
  Object {
    "payload": false,
    "type": "IS_LOADING",
  },
]
`;

exports[`testing operations doReqeuest doRequest with string, update letters 1`] = `
Array [
  Object {
    "payload": true,
    "type": "IS_LOADING",
  },
  Object {
    "payload": Object {
      "letters": Array [
        "abc",
        "def",
      ],
    },
    "type": "UPDATE_LETTERS",
  },
  Object {
    "type": "INCREMENT_PAGE",
  },
  Object {
    "payload": false,
    "type": "IS_LOADING",
  },
]
`;

exports[`testing operations doReqeuest doRequest with undefined, error expected 1`] = `
Array [
  Object {
    "payload": true,
    "type": "IS_LOADING",
  },
  Object {
    "payload": true,
    "type": "HAS_ERROR",
  },
]
`;
